File: vc_reading/summarize.py
"""Generate study-oriented takeaways for an essay using Claude.

Two paths:
* Full-text — when we successfully fetched the essay, summarize the text.
* Knowledge-based — when the source site blocked the fetch (notably Gurley's
  abovethecrowd.com on CI), ask the model to summarize the *known* canonical
  essay from the title/author, with an explicit guard: if it isn't confident it
  knows the specific piece, it returns empty so we don't fabricate takeaways.
"""

import json
import logging
import os
import re
from typing import Dict, List, Optional

from anthropic import Anthropic

logger = logging.getLogger(__name__)


def _model_chain() -> List[str]:
    """Models tried in order. Opus 4.8 first (the user's choice); fall back to a
    model the repo already calls successfully so one bad model ID can't blank
    every summary."""
    chain = []
    if os.getenv("VC_MODEL"):
        chain.append(os.getenv("VC_MODEL"))
    chain += ["claude-opus-4-8", "claude-sonnet-4-20250514"]
    seen, out = set(), []
    for m in chain:
        if m and m not in seen:
            seen.add(m)
            out.append(m)
    return out


_SHAPE = """Respond with ONLY a JSON object, no prose around it, in exactly this shape:
{{
  "one_liner": "<one sentence: the single core idea>",
  "takeaways": [
    "<3 to 5 crisp, specific takeaways; each one sentence; favor durable mental \
models and investor-relevant judgment over surface summary>"
  ],
  "investor_angle": "<one sentence on why this matters for an early-stage VC's \
decision-making or founder evaluation>"
}}"""

PROMPT_TEXT = """You are a mentor helping an aspiring elite early-stage venture \
capitalist study the canon. Below is an essay by {author}, titled "{title}".

Read it and distill it for a busy investor. """ + _SHAPE + """

ESSAY:
---
{text}
---"""

PROMPT_KNOWLEDGE = """You are a mentor helping an aspiring elite early-stage \
venture capitalist study the canon. Summarize the well-known essay "{title}" by \
{author} from your own knowledge of it.

IMPORTANT: Only do this if you are genuinely confident you know this specific \
essay. If you are not sure which essay this is, return {{"one_liner": "", \
"takeaways": [], "investor_angle": ""}} and nothing else — do NOT guess or \
fabricate.

""" + _SHAPE


def _fallback(title: str, reason: str) -> Dict:
    return {
        "one_liner": f"Classic essay: {title}.",
        "takeaways": [
            f"Auto-summary unavailable this run ({reason}) — read it directly "
            "via the link below."
        ],
        "investor_angle": "Part of the early-stage VC canon worth reading in full.",
        "ok": False,
    }


def _parse(raw: str, title: str) -> Dict:
    raw = raw.strip()
    raw = re.sub(r"^```(?:json)?|```$", "", raw, flags=re.MULTILINE).strip()
    match = re.search(r"\{.*\}", raw, re.DOTALL)
    data = json.loads(match.group(0) if match else raw)
    takeaways = data.get("takeaways") or []
    if isinstance(takeaways, str):
        takeaways = [takeaways]
    return {
        "one_liner": str(data.get("one_liner", "")).strip(),
        "takeaways": [str(t).strip() for t in takeaways if str(t).strip()],
        "investor_angle": str(data.get("investor_angle", "")).strip(),
        "ok": True,
    }


def summarize(author_name: str, title: str, text: Optional[str], url: str,
              client: Optional[Anthropic] = None) -> Dict:
    """Return {one_liner, takeaways[list], investor_angle, ok}. Never raises."""
    if not os.getenv("ANTHROPIC_API_KEY") and client is None:
        return _fallback(title, "ANTHROPIC_API_KEY not set")

    if text:
        prompt = PROMPT_TEXT.format(author=author_name, title=title, text=text)
        mode = "full-text"
    else:
        logger.warning("No fetched text for '%s' — summarizing from knowledge.", title)
        prompt = PROMPT_KNOWLEDGE.format(author=author_name, title=title)
        mode = "knowledge"

    client = client or Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
    last_err = None
    for model in _model_chain():
        try:
            resp = client.messages.create(
                model=model, max_tokens=1024,
                messages=[{"role": "user", "content": prompt}],
            )
            result = _parse(resp.content[0].text, title)
            if result["takeaways"]:
                logger.info("Summarized '%s' (%s, %s).", title, mode, model)
                return result
            last_err = "model returned empty (likely unknown essay)"
        except Exception as exc:
            last_err = str(exc)
            logger.warning("Model %s failed for '%s': %s", model, title, exc)
            continue
    logger.error("Summary failed for '%s': %s", title, last_err)
    return _fallback(title, "fetch blocked & essay not recognized"
                     if mode == "knowledge" else "model call failed")

File: skim/lib/summarize.ts
import Anthropic from "@anthropic-ai/sdk";

// Ported from vc_reading/summarize.py — same JSON shape and anti-hallucination
// guard, generalized from "investor angle" to a single "key insight" so it works
// for any newsletter or blog, not just the VC canon.

export type Summary = {
  one_liner: string;
  takeaways: string[];
  key_insight: string;
};

const SHAPE = `Respond with ONLY a JSON object, no prose around it, in exactly this shape:
{
  "one_liner": "<one sentence: the single core idea>",
  "takeaways": ["<3 to 5 crisp, specific takeaways; each one sentence; favor durable ideas and concrete specifics over generic summary>"],
  "key_insight": "<one sentence on the most useful or non-obvious thing to remember>"
}`;

function prompt(title: string, source: string, text: string): string {
  return `You distill writing for a busy, smart reader who has ~20 seconds per piece. Below is a post titled "${title}" from "${source}".

Read it and distill it. ${SHAPE}

POST:
---
${text.slice(0, 16000)}
---`;
}

function modelChain(): string[] {
  const chain = [process.env.SKIM_MODEL, "claude-haiku-4-5-20251001", "claude-sonnet-4-6"];
  return [...new Set(chain.filter(Boolean) as string[])];
}

function parse(raw: string): Summary | null {
  const start = raw.indexOf("{");
  const end = raw.lastIndexOf("}");
  if (start === -1 || end === -1) return null;
  try {
    const obj = JSON.parse(raw.slice(start, end + 1));
    const takeaways = Array.isArray(obj.takeaways)
      ? obj.takeaways.filter((t: unknown) => typeof t === "string" && t.trim()).slice(0, 5)
      : [];
    if (!obj.one_liner || takeaways.length === 0) return null;
    return {
      one_liner: String(obj.one_liner).trim(),
      takeaways,
      key_insight: String(obj.key_insight || "").trim(),
    };
  } catch {
    return null;
  }
}

export async function summarize(title: string, source: string, text: string): Promise<Summary | null> {
  if (!text || text.trim().length < 200) return null; // not enough to summarize honestly
  const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  for (const model of modelChain()) {
    try {
      const resp = await client.messages.create({
        model,
        max_tokens: 700,
        messages: [{ role: "user", content: prompt(title, source, text) }],
      });
      const block = resp.content.find((b) => b.type === "text");
      const out = block && block.type === "text" ? parse(block.text) : null;
      if (out) return out;
    } catch (err) {
      console.warn(`summarize: model ${model} failed:`, (err as Error).message);
    }
  }
  return null;
}
